### PR TITLE
Dev/update gaps script

### PR DESCRIPTION
**Pull request by mardet987**
Updating borealis_gaps script to work on array restructured files (results in a much faster processing time!), using pydarn. Some configurability updates

### CODE REVIEW

File: borealis_gaps.py [hunk 1]
```diff
@@ -4,71 +4,78 @@
...
     return usage_message
 
 
 def borealis_gaps_parser():
     parser = argparse.ArgumentParser(usage=usage_msg())
-    parser.add_argument("data_dir", help="Path to the directory that holds the day directories "
-                                         "containing *[filetype].hdf5.site.bz2 files.")
-    parser.add_argument("start_day", help="First day directory to check, given as YYYYMMDD.")
-    parser.add_argument("end_day", help="Last day directory to check, given as YYYYMMDD.")
+    parser.add_argument("data_dir", help="Path to the directory that holds any directory structure which within "
```
Comment (kevinkrieger) [inline borealis_gaps.py:60]: only works on hdf5 or site files, not dmap, correct?

File: borealis_gaps.py [hunk 2]
```diff
@@ -4,71 +4,78 @@
...
+This script must be run from a virtualenv with pydarn installed.
+
 This script should be piped to a file gaps.md or similar.
 
 This script generates text in a markdown table format. The following
-command line call can be used to generate a docx table from 
+command line call can be used to generate a docx table from
 the printed output:
 
 pandoc -o output_file.docx -f markdown -t docx input_file.md
```
Comment (kevinkrieger) [inline borealis_gaps.py:13]: nice

File: borealis_gaps.py [hunk 3]
```diff
@@ -109,37 +116,42 @@ def get_record_timestamps(filename, record_dict, file_structure='site'):
...
     if file_structure == 'site':
-        records = sorted(deepdish.io.load(borealis_hdf5_file).keys())
+        if os.path.basename(filename).split('.')[-1] in ['bz2', 'bzip2']:
+            borealis_hdf5_file = decompress_bz2(filename)
+            bzip2 = True
+        else:
+            borealis_hdf5_file = filename
+            bzip2 = False
+
+        reader = pydarn.BorealisRead(filename, filetype, file_structure)
```
Comment (kevinkrieger) [inline borealis_gaps.py:130]: Does BorealisRead ever fail?

File: borealis_gaps.py [hunk 4]
```diff
@@ -109,37 +116,42 @@ def get_record_timestamps(filename, record_dict, file_structure='site'):
...
+        # get all records first timestamp
+        records = []
+        for record_name in reader.record_names:
+            records.append(reader.records[record_name]['sqn_timestamps'][0])
+
+        if bzip2:
+            if borealis_hdf5_file.split('.')[-1] in ['bz2', 'bzip2']:
```
Comment (kevinkrieger) [inline borealis_gaps.py]: I'm not sure I understand this if statement, would this happen if the filename had two 'bz2's at the end?

File: borealis_gaps.py [hunk 5]
```diff
@@ -266,50 +278,59 @@ def print_gaps(gaps_dict, first_timestamp, last_timestamp):
...
-                duration = gap_duration.total_seconds()
-                duration_min = round(duration/60.0, 1)
-                print('| ' + gap_start_time.strftime(strf_format) + ' | '  +
-                      gap_end_time.strftime(strf_format) + ' | ' + 
-                      str(duration_min) + ' |   |')
-                duration_dict[day] = duration_min
-
-    # end table, print new line
+    with open(print_filename, "a") as f:
+        print('GAPS GREATER THAN {} s BETWEEN {} and {}:'.format(str(gap_spacing), first_timestamp.strftime(strf_format), last_timestamp.strftime(strf_format)), file=f)
```
Comment (kevinkrieger) [inline borealis_gaps.py]: long line, could split the format into multiple lines

File: borealis_gaps.py [hunk 6]
```diff
@@ -266,50 +278,59 @@ def print_gaps(gaps_dict, first_timestamp, last_timestamp):
...
+                    duration_dict[day] += duration_min
+
+        # end table, print new line
+        print(' ', file=f)
+        total_duration_min = 0.0
+        for day, duration in duration_dict.items():
+            total_duration_min += duration
+        total_duration_hrs = round(total_duration_min/60.0, 1)
+        total_duration_days = round(total_duration_hrs/24.0, 1)
+        print('TOTAL DOWNTIME DURATION IN PERIOD from {} to {}: \n'.format(first_timestamp.strftime(strf_format), last_timestamp.strftime(strf_format)), file=f)
```
Comment (kevinkrieger) [inline borealis_gaps.py]: long line

File: borealis_gaps.py [hunk 7]
```diff
@@ -349,10 +379,13 @@ def print_gaps(gaps_dict, first_timestamp, last_timestamp):
     for one_day in daterange(start_day, end_day):
         # Get all the filenames and then all the timestamps for this day.
         print(one_day.strftime("%Y%m%d"))
-        if os.path.isdir(data_dir + one_day.strftime("%Y%m%d")):
-            files = glob.glob(data_dir + one_day.strftime("%Y%m%d") + '/*.' + filetype + '*')
-        else:
-            continue
+
+        files = subprocess.check_output(['find', data_dir, '-name',
+                    one_day.strftime("%Y%m%d")+'*.' + filetype + '*' + file_extension]).splitlines()
```
Comment (kevinkrieger) [inline borealis_gaps.py]: does this work without single quotes around the argument to -name ?

File: borealis_gaps.py [hunk 8]
```diff
@@ -364,14 +397,15 @@ def print_gaps(gaps_dict, first_timestamp, last_timestamp):
         while files_left:
             for procnum in range(num_processes):
                 try:
-                    filename = files[filename_index + procnum]
+                    filename = files[filename_index + procnum].decode('ascii')
                 except IndexError:
                     if filename_index + procnum == 0:
-                        print('No files found to check!')
-                        raise
+                        print('No files found for this day!')
```
Comment (kevinkrieger) [inline borealis_gaps.py]: Could print the date instead of "this day"

File: borealis_gaps.py [hunk 9]
```diff
@@ -381,20 +415,34 @@ def print_gaps(gaps_dict, first_timestamp, last_timestamp):
 
             filename_index += num_processes
 
-        record_dict[one_day] = filename_dict
+        if files_this_day:
+            record_dict[one_day] = filename_dict
+            timestamps_dict[one_day] = combine_timestamp_lists(record_dict[one_day])
+        
+        if one_day == start_day:
```
Comment (kevinkrieger) [inline borealis_gaps.py:426]: Handle first timestamp and last timestamp in if/elif statement

File: borealis_gaps.py (hunk 4 above)
Comment (kevinkrieger) [reply borealis_gaps.py]: I think I see now, you're protecting the original. I guess can you error-check the decompress_bz2?

**Review by kevinkrieger — APPROVED** [review]
a few minor comments.

File: borealis_gaps.py (hunk 4 above)
Comment (mardet987) [reply borealis_gaps.py]: I think I was more afraid that bzip2 could get set unintentionally, if this code got modified or something. I modified this a little.

File: borealis_gaps.py (hunk 3 above)
Comment (mardet987) [reply borealis_gaps.py:130]: yes, I think it's preferable to just raise though

**Pull request merged**